lib/rules/no-extra-boolean-cast.js

```diff
@@ -115,12 +121,49 @@ module.exports = {
                 return isInFlaggedContext(node.parent);
             }
 
-            return isInBooleanContext(node) ||
-            (isLogicalContext(node.parent) &&
+            /*
+             * legacy behavior - enforceForLogicalOperands will only recurse on
+             * logical expressions, not on other contexts.
+             * enforceForInnerExpressions will recurse on the rest.
+             */
 
-            // For nested logical statements
-            isInFlaggedContext(node.parent)
-            );
+            if (enforceForLogicalOperands || enforceForInnerExpressions) {
+                if (isLogicalContext(node.parent)) {
```

Comment:
I think this whole function would be easier to follow if we now remove `function isLogicalContext` (which was a bit confusing name anyway) and move the condition here to align this check with the new checks below.